agent/system_metrics.go

```diff
@@ -0,0 +1,226 @@
+	Format() DataFormat
+	// Collect returns the value of the collected metric when the function is called.
+	Collect() ([]byte, error)
+}
+
+type SystemMetricCollector struct {
+	mu          sync.Mutex
+	wg          sync.WaitGroup
+	innerCancel context.CancelFunc
+	logger      client.LoggerProducer
+	taskOpts    *system_metrics.SystemMetricsOptions
+	connOpts    system_metrics.ConnectionOptions
+	interval    time.Duration
+	collectors  []MetricCollector
+	id          string
+	client      *system_metrics.SystemMetricsClient
+	errored     bool
+	closed      bool
+}
+
+// DataFormat describes how the time series data is stored.
+type DataFormat int32
+
+// Valid DataFormat values.
+const (
+	DataFormatText DataFormat = 0
+	DataFormatFTDC DataFormat = 1
+	DataFormatBSON DataFormat = 2
+	DataFormatJSON DataFormat = 3
+	DataFormatCSV  DataFormat = 4
+)
+
+type diskUsageCollector struct{}
+
+// NewDiskUsageCollector creates a diskUsageCollector object.
+func NewDiskUsageCollector() *diskUsageCollector {
+	return new(diskUsageCollector)
+}
+
+func (collector *diskUsageCollector) Name() string { return "disk_usage" }
+
+func (collector *diskUsageCollector) Format() DataFormat { return DataFormatFTDC }
+
+func (collector *diskUsageCollector) Collect(ctx context.Context, dir string) ([]byte, error) {
+	usage, err := disk.Usage(dir)
+	if err != nil {
+		return nil, errors.Wrap(err, "problem capturing metrics with gopsutil")
+	}
+
+	return convertJSONToFTDC(ctx, usage)
+}
+
+func convertJSONToFTDC(ctx context.Context, metric interface{}) ([]byte, error) {
+	jsonMetrics, err := json.Marshal(metric)
+	if err != nil {
+		return nil, errors.Wrap(err, "problem converting metrics to JSON")
+	}
+
+	opts := metrics.CollectJSONOptions{
+		InputSource:   bytes.NewReader(jsonMetrics),
+		SampleCount:   1,
+		FlushInterval: 1 * time.Second,
+	}
+
+	output, err := metrics.CollectJSONStream(ctx, opts)
+	if err != nil {
+		return nil, errors.Wrap(err, "problem converting FTDC to JSON")
+	}
+	return output, nil
+}
+
+type uptimeCollector struct{}
+
+// NewUptimeCollector creates an uptimeCollector object.
+func NewUptimeCollector() *uptimeCollector {
+	return new(uptimeCollector)
+}
+
+type uptimeWrapper struct {
+	Uptime uint64 `json:"uptime"`
+}
+
+func (collector *uptimeCollector) Name() string { return "uptime" }
+
+func (collector *uptimeCollector) Format() DataFormat { return DataFormatFTDC }
+
+func (collector *uptimeCollector) Collect(ctx context.Context) ([]byte, error) {
+	uptime, err := host.Uptime()
+	if err != nil {
+		return nil, errors.Wrap(err, "problem capturing metrics with gopsutil")
+	}
+
+	uptimeWrap := uptimeWrapper{uptime}
+	return convertJSONToFTDC(ctx, uptimeWrap)
+}
+
+type processCollector struct{}
+
+// NewProcessCollector creates a processCollector object.
+func NewProcessCollector() *processCollector {
+	return new(processCollector)
+}
+
+type processesWrapper struct {
+	Processes []ProcessData `json:"processes"`
+}
+
+type ProcessData struct {
```

Comment:
Do we need to export this? If so, add a comment.

Reply:
It shouldn't need to be exported.

Reply:
changed to unexported